### PR TITLE
feat: lower Influx regex expressions to DF regex expressions 

### PR DESCRIPTION
**Pull request by crepererum**
~On top of #6389.~

For #6388.

### CODE REVIEW

File: query_tests/cases/in/pushdown.expected [hunk 1]
```diff
@@ -320,18 +320,18 @@
...
-|               |   Filter: influx_regex_match(CAST(restaurant.town AS Utf8)restaurant.town, Utf8("foo|bar|baz")) AND influx_regex_not_match(CAST(restaurant.town AS Utf8)restaurant.town, Utf8("one|two"))                                                                                                        |
-|               |     Projection: CAST(restaurant.town AS Utf8) AS CAST(restaurant.town AS Utf8)restaurant.town, restaurant.count, restaurant.system, restaurant.time, restaurant.town                                                                                                                             |
-|               |       TableScan: restaurant projection=[count, system, time, town], partial_filters=[influx_regex_match(CAST(restaurant.town AS Utf8), Utf8("foo|bar|baz")), influx_regex_not_match(CAST(restaurant.town AS Utf8), Utf8("one|two"))]                                                             |
```
[the remark this answers is not in the document]
Comment (crepererum) [reply query_tests/cases/in/pushdown.expected:329]: https://github.com/apache/arrow-datafusion/pull/4614

File: iox_query/src/logical_optimizer/influx_regex_to_datafusion_regex.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        expr_rewriter::ExprRewriter, utils::from_plan, BinaryExpr, LogicalPlan, Operator,
+    },
+    optimizer::{utils::rewrite_preserving_name, OptimizerConfig, OptimizerRule},
+    prelude::Expr,
+    scalar::ScalarValue,
+};
+use query_functions::{clean_non_meta_escapes, REGEX_MATCH_UDF_NAME, REGEX_NOT_MATCH_UDF_NAME};
+
+#[derive(Debug, Clone)]
+pub struct InfluxRegexToDataFusionRegex {}
```
Comment (alamb) [inline iox_query/src/logical_optimizer/influx_regex_to_datafusion_regex.rs:14]: I think some comments on the rationale for this optimization would help here

Specifically, that the datafusion regex implementation is faster / more optimized than our IOx one so
it should be used if/when possible

**Review by alamb — APPROVED** [review]

**Comment by crepererum** [comment]
Not ready, blocked by linked DF ticket.

**Comment by alamb** [comment]
(FWIW I think this PR could be merged as is, although https://github.com/apache/arrow-datafusion/pull/4614 would make it even better)

**Comment by alamb** [comment]
PR to upgrade DataFusion https://github.com/influxdata/influxdb_iox/pull/6407

File: iox_query/src/logical_optimizer/influx_regex_to_datafusion_regex.rs (hunk 2 above)
Comment (crepererum) [reply iox_query/src/logical_optimizer/influx_regex_to_datafusion_regex.rs:18]: done

File: iox_query/src/logical_optimizer/influx_regex_to_datafusion_regex.rs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                        let op = match fun.name.as_str() {
+                            REGEX_MATCH_UDF_NAME => Operator::RegexMatch,
+                            REGEX_NOT_MATCH_UDF_NAME => Operator::RegexNotMatch,
+                            _ => unreachable!(),
+                        };
+                        return Ok(Expr::BinaryExpr(BinaryExpr {
+                            left: Box::new(args.remove(0)),
+                            op,
+                            right: Box::new(Expr::Literal(ScalarValue::Utf8(Some(s)))),
+                        }));
```
[the remark this answers is not in the document]
Comment (crepererum) [reply iox_query/src/logical_optimizer/influx_regex_to_datafusion_regex.rs]: done

File: iox_query/src/logical_optimizer/mod.rs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+use std::sync::Arc;
+
+use datafusion::optimizer::{optimizer::Optimizer, OptimizerConfig};
+
+use self::influx_regex_to_datafusion_regex::InfluxRegexToDataFusionRegex;
+
+mod influx_regex_to_datafusion_regex;
+
+pub fn iox_optimizer() -> Optimizer {
```
[the remark this answers is not in the document]
Comment (crepererum) [reply iox_query/src/logical_optimizer/mod.rs:12]: done

File: query_tests/cases/in/pushdown.expected (hunk 1 above)
[the remark this answers is not in the document]
Comment (crepererum) [reply query_tests/cases/in/pushdown.expected:329]: done

**Pull request merged**